// react-sextant/lib/root-view/progress.js
import React, {Component} from "react";
import {
    StyleSheet,
    AppRegistry,
    View,
    Platform,
    ProgressViewIOS,
    TouchableOpacity,
    Text,
    DeviceEventEmitter
} from 'react-native';
import * as Progress from 'react-native-progress';

let viewRoot = null;

class RootView extends Component {
    constructor(props) {
        super(props);
        viewRoot = this;
        this.state = {
            toast: null,
            progressNum:0,
        }
    }

    render() {
        return (
            <React.Fragment>
                <View style={styles.rootView} pointerEvents="box-none" zIndex={999}>
                    {!!this.state.toast?this.state.toast:(this.state.progressNum>0&&
                        (Platform.OS==='android'?
                            <React.Fragment>
                                <Progress.Circle
                                    style={styles.progress}
                                    progress={this.state.progressNum}
                                    size={84}
                                    unfilledColor="rgba(255,255,255,0.5)"
                                    color={"#BA9866"}
                                    thickness={6}
                                    showsText={true}
                                    textStyle={{fontSize:20,color:'#BA9866',alignSelf:'center'}}
                                />
                                <View style={styles.progress}>
                                    <TouchableOpacity onPress={()=>{
                                        if(DeviceEventEmitter.listeners("fetch_download").length>0){
                                            DeviceEventEmitter.emit("fetch_download");
                                        }
                                        this.setState({progressNum:0,toast:null})
                                    }}>
                                        <Text style={{marginTop:130,padding:10,textDecorationLine:'underline'}}>取消下载</Text>
                                    </TouchableOpacity>
                                </View>
                            </React.Fragment>:
                            <ProgressViewIOS progress={this.state.progressNum}/>))
                    }
                </View>

            </React.Fragment>
        )
    }

    static setView = (view) => {
        viewRoot.setState({toast: view})
    };

    static setLoading = (progress) => {
        viewRoot.setState({progressNum: progress})
    }
}


const originRegister = AppRegistry.registerComponent;

AppRegistry.registerComponent = (appKey, component) => {

    return originRegister(appKey, function () {
        const OriginAppComponent = component();

        return class extends Component {
            render() {
                return (
                    <View style={styles.container}
                          onStartShouldSetResponder={()=>RootView.setView(null)}
                    >
                        <View style={{flex:1}} zIndex={1}>
                            <OriginAppComponent/>
                        </View>
                        <RootView/>
                    </View>
                );
            };
        };
    });
};
const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor:'transparent',
        position: 'relative',
    },
    rootView: {
        position: "absolute",
        left: 0,
        right: 0,
        top: 0,
        bottom: 0,
        flexDirection: "row",
        justifyContent: "center",
    },
    progress: {
        position: 'absolute',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        alignItems: 'center',
        justifyContent: 'center',
        zIndex: 999,
    },
});
export default RootView
